raiden/raiden_service.py

```diff
@@ -635,6 +651,214 @@ def settle(self, token_address, partner_address):
+                while any(c.state != 'settled' for c in open_channels):
+                    gevent.sleep(1)
+                    if time.time() > timeout:
+                        log.debug(
+                            "timeout while waiting for settlement",
+                            unsettled=len(filter(lambda c: c.state != 'settled', open_channels)),
+                            settled=len(filter(lambda c: c.state == 'settled', open_channels))
+                        )
+                        break
+
+    def _join_channel(self, partner_address, partner_deposit):
+        """Will be called, when we were selected as channel partner by another
+        node. It will fund the channel with up to the partner's deposit, but not
+        more than remaining funds or the initial funding per channel.
+
+        If the connection manager has no funds, this is a noop.
+        """
+        # not initialized
+        if self.funds <= 0:
+            return
+        with self.lock:
+            remaining = self.funds_remaining
+            initial = self.initial_funding_per_partner
+            joining_funds = min(
+                partner_deposit,
+                remaining,
+                initial
+            )
+            if joining_funds <= 0:
+                return
+
+            self.raiden.api.deposit(
+                self.token_address.encode('hex'),
+                partner_address.encode('hex'),
+                joining_funds
+            )
+            log.debug(
+                "joined a channel!",
+                funds=joining_funds,
+                me=pex(self.raiden.address),
+                partner=pex(partner_address)
+            )
+
+    def _retry_connect(self):
+        """Will be called when new channels in the token network are detected.
+        If the minimum number of channels was not yet established, it will try
+        to open new channels.
+
+        If the connection manager has no funds, this is a noop.
+        """
+        # not initialized
+        if self.funds <= 0:
+            return
+        with self.lock:
+            if self.funds_remaining <= 0:
+                return
+            if len(self.open_channels) >= self.initial_channel_target:
+                return
+            for partner in self.find_new_partners(self.initial_channel_target - len(self.open_channels)):
+                self.raiden.api.open(
+                    self.token_address.encode('hex'),
+                    partner.encode('hex')
+                )
+                self.raiden.api.deposit(
+                    self.token_address.encode('hex'),
+                    partner.encode('hex'),
+                    self.initial_funding_per_partner
+                )
+
+    def find_new_partners(self, number):
+        """Search the token network for potential channel partners.
+        Args:
+            number (int): number of partners to return
+        """
+        known = set([c.partner_address for c in self.open_channels])
```

Comment:
We don't need the temporary list, `set(channel.partner_address for channel in self.open_channels)` is sufficient.